temporalio/worker/_workflow.py

```diff
@@ -253,41 +263,44 @@ async def _handle_activation(
             if workflow:
                 activate_task = asyncio.get_running_loop().run_in_executor(
                     self._workflow_task_executor,
-                    workflow.activate,
+                    workflow.instance.activate,
                     act,
                 )
 
-                # Wait for deadlock timeout and set commands if successful
+                # Run activation task with deadlock timeout
                 try:
                     completion = await asyncio.wait_for(
                         activate_task, self._deadlock_timeout_seconds
                     )
                 except asyncio.TimeoutError:
-                    raise _DeadlockError.from_deadlocked_workflow(
-                        workflow, self._deadlock_timeout_seconds
-                    ) from None
+                    # Need to create the deadlock exception up here so it
+                    # captures the trace now instead of later after we may have
+                    # interrupted it
+                    deadlock_exc = _DeadlockError.from_deadlocked_workflow(
+                        workflow.instance, self._deadlock_timeout_seconds
+                    )
+                    # When we deadlock, we will raise an exception to fail
+                    # the task. But before we do that, we want to try to
+                    # interrupt the thread and put this activation task on
+                    # the workflow so that the successive eviction can wait
+                    # on it before trying to evict.
+                    deadlocked_thread_id = workflow.instance.get_thread_id()
+                    if deadlocked_thread_id:
+                        temporalio.bridge.runtime.Runtime._raise_in_thread(
+                            deadlocked_thread_id, _InterruptDeadlockError
+                        )
```

Comment:
The timed-out task could have completed between line 275 and line 289. In that case, the event loop in the thread that it was running on may now be processing a task from an unrelated workflow, and we will throw an exception in that workflow. Is that what you were referring to here?

>  It technically could cause user issues if they were expecting their Python not to be interrupted even after deadlock timeout, but we don't support code running successfully after deadlock timeout.

Reply:
> In that case, the event loop in the thread that it was running on may now be processing a task from an unrelated workflow

This thread is only given back to the thread pool for reuse when no more code is run on it (i.e. when all of our workflow instance code reached some stopping point). And only we control the event loop for successive tasks and we don't trigger that until eviction after this task is confirmed done.

> Is that what you were referring to here?

Kinda. I am saying technically if you had:

```python
try:
    while True:
        pass
finally:
    nuke_the_world()
```

We would never have run `nuke_the_world()` before because we had no concept of interrupting a deadlock, but we would run it now. But we don't ensure compatibility for any code post deadlock.

Reply:
> In that case, the event loop in the thread that it was running on may now be processing a task from an unrelated workflow

Hrmm, this is true. I wonder if there is a reasonable way without too much effort to prevent the thread from being returned to the pool if we encounter a deadlock. Hopefully it doesn't require a mutex or us to hang the activate call open on some thread-blocking thing if we see deadlock. I will think on it. Any ideas? If it's too complicated we may have to scrap the idea of interrupting the deadlock this way.

> Is that what you were referring to here?

No, this was something else (technically we interrupt their deadlock where we didn't before, possibly causing other code to run)